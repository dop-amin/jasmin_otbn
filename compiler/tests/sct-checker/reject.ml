open Jasmin
open Common
open Sct_checker_forward

let path = "fail"

let load_and_check name =
  Jasmin.Utils.nowarning ();
  Format.printf "File %s:@." name;
  let ((_, fds) as p) = load_file (Filename.concat path name) in
  List.iter
    (fun fd ->
      if should_check fd then
        let f = fd.Prog.f_name.fn_name in
        match ty_prog Arch.is_ct_sopn p [ f ] with
        | exception Utils.HiError e ->
            let e = { e with err_loc = Lnone } in
            Format.printf "Failed as expected %s: %a@." f Utils.pp_hierror e
        | _ ->
            Format.eprintf "Did not fail: %s.@." f;
            assert false)
    fds

let () =
  let cases = Sys.readdir path in
  Array.sort String.compare cases;
  Array.iter load_and_check cases
