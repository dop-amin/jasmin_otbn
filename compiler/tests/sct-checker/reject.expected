File basic.jazz:
Failed as expected xof_init: speculative constant type checker: return type for state is #[ptr = public, val = secret] it should be less than #[ptr = public, val = { n = public, s = t}]
Failed as expected missing_then: speculative constant type checker: the variable m is not known to be a msf, only {
                                   } are
Failed as expected missing_else: speculative constant type checker: the variable m is not known to be a msf, only {
                                   } are
Failed as expected after_branch: speculative constant type checker: the variable m is not known to be a msf, only {
                                   } are
Failed as expected leak_transient: speculative constant type checker: x has type #transient but should be at most #public
File functions.jazz:
Failed as expected return_outdated: speculative constant type checker: the variable msf is not known to be a msf, only {
                                      } are
Failed as expected return_unknown: speculative constant type checker: the variable msf is not known to be a msf, only {
                                     } are
Failed as expected call_kills_msf: speculative constant type checker: 
                                   this function call destroys MSFs and { m } are required.
                                   Trace:
                                     the function leak destroys MSFs at "fail/functions.jazz", line 2 (36-48)
File inline-if-modmsf.jazz:
Failed as expected inline_if_modmsf: speculative constant type checker: 
                                     this function call destroys MSFs and { ms } are required.
                                     Trace:
                                       the function f destroys MSFs at "fail/inline-if-modmsf.jazz", line 6 (8) to line 8 (9)
File modmsf-trace.jazz:
Failed as expected modmsf_trace: speculative constant type checker: this function call destroys MSFs and { ms } are required.
                                 Trace:
                                   the function f2 destroys MSFs at "fail/modmsf-trace.jazz", line 13 (19-25)
                                   the function f1 destroys MSFs at "fail/modmsf-trace.jazz", line 10 (19-31)
File rsb.jazz:
Failed as expected pointed_data_is_transient: speculative constant type checker: x has type #transient but should be at most #public
Failed as expected inline_int_leaks: speculative constant type checker: i has type #transient but should be at most #public
Failed as expected inline_bool_leaks: speculative constant type checker: b has type #transient but should be at most #public
Failed as expected inline_modmsf: speculative constant type checker: this function is annotated as nomodmsf, but it modifies the MSF at 
                                  "fail/rsb.jazz", line 295 (4-22).
                                  Trace:
                                    the function call_do_nothing destroys MSFs at "fail/rsb.jazz", line 19 (4-17)
Failed as expected call_modmsf: speculative constant type checker: this function is annotated as nomodmsf, but it modifies the MSF at 
                                "fail/rsb.jazz", line 289 (4-17).
                                
Failed as expected inline_kills_msf2: speculative constant type checker: 
                                      this function call destroys MSFs and { msf } are required.
                                      Trace:
                                        the function call_do_nothing destroys MSFs at "fail/rsb.jazz", line 19 (4-17)
Failed as expected inline_kills_msf: speculative constant type checker: 
                                     this function call destroys MSFs and { msf } are required.
                                     Trace:
                                       the function inline_call_do_nothing destroys MSFs at "fail/rsb.jazz", line 25 (4-17)
Failed as expected update_after_call_many: speculative constant type checker: This function is annotated with update_after_call, but it returns more than one MSF
Failed as expected invalid_update_after_call: speculative constant type checker: This function is annotated with update_after_call, but it does not return an MSF
Failed as expected write_mmx_call_noupdate: speculative constant type checker: Assignment of type #transient to m not allowed. MMX registers must always be public.
Failed as expected write_mmx_call: speculative constant type checker: Assignment of type #transient to m not allowed. MMX registers must always be public.
Failed as expected write_mmx_ptr: speculative constant type checker: Assignment of type #transient to m not allowed. MMX registers must always be public.
Failed as expected write_mmx_opn: speculative constant type checker: Assignment of type #transient to m not allowed. MMX registers must always be public.
Failed as expected write_mmx_assign2: speculative constant type checker: Assignment of type #secret to m not allowed. MMX registers must always be public.
Failed as expected write_mmx_assign: speculative constant type checker: Assignment of type #transient to m not allowed. MMX registers must always be public.
Failed as expected call_in_for_noupdate: speculative constant type checker: return type for x is #transient it should be less than #public
Failed as expected call_in_for: speculative constant type checker: return type for x is #transient it should be less than #public
Failed as expected call_in_while_noupdate: speculative constant type checker: msf is 
                                           Exact {  } it should be be at least 
                                           Exact { msf }
Failed as expected call_in_while_unknown: speculative constant type checker: constraints caused by the loop cannot be satisfied
Failed as expected call_in_while: speculative constant type checker: constraints caused by the loop cannot be satisfied
Failed as expected transient_after_call_noupdate2: speculative constant type checker: return type for x is #transient it should be less than #public
Failed as expected transient_after_call_noupdate1: speculative constant type checker: return type for x is #transient it should be less than #public
Failed as expected transient_after_call_noupdate: speculative constant type checker: return type for x is #transient it should be less than #public
Failed as expected transient_after_call_update: speculative constant type checker: L-value _ /* u64 */ must be a reg
Failed as expected archetypical: speculative constant type checker: x has type #transient but should be at most #public
